libvast/vast/segment.hpp

```diff
@@ -0,0 +1,192 @@
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include <caf/actor_system.hpp>
+#include <caf/expected.hpp>
+#include <caf/intrusive_ptr.hpp>
+#include <caf/ref_counted.hpp>
+#include <caf/stream_serializer.hpp>
+#include <caf/streambuf.hpp>
+
+#include "vast/aliases.hpp"
+#include "vast/chunk.hpp"
+#include "vast/fwd.hpp"
+#include "vast/uuid.hpp"
+
+namespace vast {
+
+// TODO: use a format that's more conducive to mmap'ing, e.g., flatbuffers.
+// Right now we use a packed struct for the header as a poor-man's abstraction
+// for this, but it's inconvenient.
+
+class segment;
+class segment_builder;
+
+/// @relates segment
+using segment_ptr = caf::intrusive_ptr<segment>;
+
+/// A sequence of [@ref table_slice](table slices) optimized for persistent
+/// storage. The layout has the following format:
+///
+///               +--------------------+--------------------+
+///               |       magic        |      version       | \
+///               +--------------------+--------------------+ |
+///               |                 segment                 | | fixed size
+///               |                  UUID                   | /
+///               +-----------------------------------------+ |
+///            +--|             table slice offset          | /
+///            |  +-----------------------------------------+
+///            |  .                                         . \
+///            |  .                meta data                . | variable size
+///            |  .                                         . /
+///            |  +-----------------------------------------+
+///            +->.                                         . \
+///               .                                         . |
+///               .               table slices              . | variable size
+///               .                                         . |
+///               .                                         . /
+///               +-----------------------------------------+
+///
+class segment : public caf::ref_counted {
+  friend segment_builder;
+
+public:
+  using magic_type = uint32_t;
+  using version_type = uint32_t;
+
+  /// A magic constant that identifies segment files. The four bytes represent
+  /// the multiplication of the vector `(1, 2, 3, 4)` with the value `42`,
+  /// converted to hex bytes: `42 * (1, 2, 3, 4) = [2a, 65, 7e, a8]`.
+  static inline constexpr magic_type magic = 0x2a547ea8;
+
+  /// The current version of the segment format.
+  static inline constexpr version_type version = 1;
+
+  /// The fixed-size header for every segment.
+  struct header {
+    magic_type magic;        ///< Magic constant to identify segments.
+    version_type version;    ///< Version of the segment format.
+    uuid id;                 ///< The UUID of the segment.
+    uint64_t payload_offset; ///< The offset to the table slices.
+  } __attribute__((packed));
```

Comment:
I think we should be careful to use non-portable attributes in the code base. At least we could define `VAST_PACKED` in the config header and use that instead. This at least gives us some control and we can potentially support other compiler further down the road.

Reply:
Yeah, I'm not a fan of these either. There's a big todo at the top of the header as well. I'd actually like to use flatbuffers here. Then we could also properly "serialize" the subsequent slice meta data.